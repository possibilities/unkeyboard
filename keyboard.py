import math
import cadquery as cq
from timer import timer
from fuse_parts import fuse_parts
from cq_workplane_plugin import cq_workplane_plugin
from explode_parts import explode_parts
from types import SimpleNamespace


# Defaults to an Atreus 64, with a wide bezel, and chicago bolts
default_config = SimpleNamespace(
    **{
        # Configurable
        "has_thicc_spacer": False,
        "use_chicago_bolt": True,
        "has_double_inner_keys": False,
        "angle": 10,
        "number_of_rows": 5,
        "number_of_columns": 6,
        "stagger_percent_for_single_inner_key": 8.5,
        "stagger_percent_for_double_inner_keys": 3.98,
        "column_stagger_percents": (-1, 4, 10, 5, 2, 2),
        # Structural
        "base_layer_thickness": 3,
        "inner_frame_size": 2.1,
        "outer_frame_size_for_chicago_bolt": 20,
        "outer_frame_size_for_regular_screw": 16,
        "screw_hole_radius_for_chicago_bolt": 2.5,
        "screw_hole_radius_for_regular_screw": 1.5,
        "switch_plate_key_cutout_size": 13.97,
        "distance_between_switch_centers": 19,
        "usb_cutout_width": 4,
        "top_inside_screw_distance_from_usb": 5.50,
    }
)

# View options

explode_by = 20
flatten_items = False


def flatten_list(list):
    return [item for sublist in list for item in sublist]


def find_rectangle_coords_around_point(point, width, height):
    offset = cq.Vector(*point)
    points = [
        cq.Vector(width / -2.0, height / -2.0, 0),
        cq.Vector(width / 2.0, height / -2.0, 0),
        cq.Vector(width / 2.0, height / 2.0, 0),
        cq.Vector(width / -2.0, height / 2.0, 0),
    ]
    return [((point + offset).x, (point + offset).y) for point in points]


def find_point_for_angle(vertice, distance, angel):
    angle_radian = math.pi / 2 - math.radians(angel)
    return (
        vertice[0] + distance * math.cos(angle_radian),
        vertice[1] + distance * math.sin(angle_radian),
    )


@cq_workplane_plugin
def mirror_layer(self, geometry):
    return self.mirror(
        mirrorPlane="YZ",
        union=True,
        basePointVector=(geometry.mirror_base_point),
    )


@cq_workplane_plugin
def drill_holes(part, geometry, config):
    screw_hole_radius = (
        config.screw_hole_radius_for_chicago_bolt
        if config.use_chicago_bolt
        else config.screw_hole_radius_for_regular_screw
    )
    return (
        part.pushPoints(geometry.screws)
        .circle(screw_hole_radius)
        .cutBlind(geometry.thickness)
    )


@cq_workplane_plugin
def drill_reset_button_hole(part, geometry):
    return (
        part.moveTo(*geometry.reset_button.point)
        .circle(geometry.reset_button.radius)
        .cutBlind(geometry.reset_button.thickness)
    )


def rotate(origin, point, angle):
    angle_radian = math.radians(angle)
    [origin_x, origin_y] = origin
    [point_x, point_y] = point

    rotated_x = (
        origin_x
        + math.cos(angle_radian) * (point_x - origin_x)
        - math.sin(angle_radian) * (point_y - origin_y)
    )
    rotated_y = (
        origin_y
        + math.sin(angle_radian) * (point_x - origin_x)
        + math.cos(angle_radian) * (point_y - origin_y)
    )
    return (rotated_x, rotated_y)


def rotate_about_center_of_plane(key, angle):
    return rotate((0, 0), key, angle)


def has_reached_end_of_inner_keys_row(column, row, config):
    if column == 0:
        if row > (1 if config.has_double_inner_keys else 0):
            return True
    return False


def calculate_key_positions(config):
    key_positions = []

    inner_keys_stagger_percent = (
        config.stagger_percent_for_double_inner_keys
        if config.has_double_inner_keys
        else config.stagger_percent_for_single_inner_key
    )

    number_of_columns_including_inner_keys = config.number_of_columns + 1

    for column in range(number_of_columns_including_inner_keys):
        key_positions.append([])
        for row in range(config.number_of_rows):
            if has_reached_end_of_inner_keys_row(column, row, config):
                continue

            row_offset = config.distance_between_switch_centers * row
            column_offset = config.distance_between_switch_centers * column

            column_stagger_size = (
                inner_keys_stagger_percent
                if column == 0
                else config.column_stagger_percents[column - 1]
            ) / config.distance_between_switch_centers

            stagger_offset = (
                config.distance_between_switch_centers * column_stagger_size
            )

            key_position_x = column_offset + (
                config.distance_between_switch_centers / 2
            )

            key_position_y = (
                (config.distance_between_switch_centers / 2)
                + row_offset
                + stagger_offset
            )

            key_positions[-1].append((key_position_x, key_position_y))

    return key_positions


def calculate_switch_cutout_points(key_positions, config):
    switch_cutout_points = []

    number_of_columns_including_inner_keys = config.number_of_columns + 1

    for column in range(number_of_columns_including_inner_keys):
        switch_cutout_points.append([])
        for row in range(config.number_of_rows):
            if has_reached_end_of_inner_keys_row(column, row, config):
                continue

            key_position = key_positions[column][row]

            switch_cutout_corner_points = find_rectangle_coords_around_point(
                key_position,
                config.switch_plate_key_cutout_size,
                config.switch_plate_key_cutout_size,
            )

            switch_cutout_points[-1].append(
                [
                    rotate_about_center_of_plane(cutout, config.angle)
                    for cutout in switch_cutout_corner_points
                ]
            )

    return flatten_list(switch_cutout_points)


def calculate_switch_outline_points(key_positions, config):
    widen_cutout_around_key_size = 1

    widen_cutout_around_inner_keys_size = (
        0 if config.has_double_inner_keys else 1.5
    )
    inner_keys_unit_height = 1 if config.has_double_inner_keys else 1.5

    inner_keys_height = (
        config.distance_between_switch_centers * inner_keys_unit_height
    ) + widen_cutout_around_inner_keys_size

    outline_size_per_key = (
        config.distance_between_switch_centers + widen_cutout_around_key_size
    )

    inner_key_padding_y = (
        inner_keys_height - config.distance_between_switch_centers
    ) / 2
    outline_size = outline_size_per_key / 2
    top_left_key = (
        key_positions[0][-1][0],
        key_positions[0][-1][1] + inner_key_padding_y,
    )
    top_right_key = key_positions[-1][-1]
    bottom_left_key = (
        key_positions[0][0][0],
        key_positions[0][0][1] - inner_key_padding_y,
    )
    bottom_right_key = key_positions[-1][0]

    top_row = [column[-1] for column in key_positions]
    bottom_row = [column[0] for column in key_positions]

    top_row_points = []
    for index, key_position in enumerate(top_row):
        if index != len(top_row) - 1:
            next_key_position = top_row[index + 1]
            is_lower_than_next = next_key_position[1] > key_position[1]
            next_key_has_same_stagger = next_key_position[1] == key_position[1]
            if not next_key_has_same_stagger:
                inner_key_padding_y = (
                    (
                        inner_keys_height
                        - config.distance_between_switch_centers
                        - widen_cutout_around_key_size
                    )
                    / 2
                    if index == 0
                    else 0
                )

                left_point = (
                    (
                        next_key_position[0] - outline_size,
                        key_position[1] + outline_size + inner_key_padding_y,
                    )
                    if is_lower_than_next
                    else (
                        key_position[0] + outline_size,
                        key_position[1] + outline_size + inner_key_padding_y,
                    )
                )

                right_point = (
                    (
                        next_key_position[0] - outline_size,
                        next_key_position[1] + outline_size,
                    )
                    if is_lower_than_next
                    else (
                        key_position[0] + outline_size,
                        next_key_position[1] + outline_size,
                    )
                )

                top_row_points.insert(0, left_point)
                top_row_points.insert(0, right_point)

    bottom_row_points = []
    for index, key_position in enumerate(bottom_row):
        if index != len(bottom_row) - 1:
            next_key_position = bottom_row[index + 1]
            is_lower_than_next = next_key_position[1] > key_position[1]
            next_key_has_same_stagger = next_key_position[1] == key_position[1]
            if not next_key_has_same_stagger:
                inner_key_padding_y = (
                    (
                        inner_keys_height
                        - config.distance_between_switch_centers
                        - widen_cutout_around_key_size
                    )
                    / 2
                    if index == 0
                    else 0
                )

                left_point = (
                    (
                        key_position[0] + outline_size,
                        key_position[1] - outline_size - inner_key_padding_y,
                    )
                    if is_lower_than_next
                    else (
                        next_key_position[0] - outline_size,
                        key_position[1] - outline_size - inner_key_padding_y,
                    )
                )

                right_point = (
                    (
                        key_position[0] + outline_size,
                        next_key_position[1] - outline_size,
                    )
                    if is_lower_than_next
                    else (
                        next_key_position[0] - outline_size,
                        next_key_position[1] - outline_size,
                    )
                )

                bottom_row_points.append(left_point)
                bottom_row_points.append(right_point)

    inner_keys_height = (
        config.distance_between_switch_centers * inner_keys_unit_height
    ) + widen_cutout_around_inner_keys_size

    bottom_left_corner = (
        bottom_left_key[0] - outline_size,
        bottom_left_key[1] - outline_size + (widen_cutout_around_key_size / 2),
    )
    bottom_right_corner = (
        bottom_right_key[0] + outline_size,
        bottom_right_key[1] - outline_size,
    )
    top_right_corner = (
        top_right_key[0] + outline_size,
        top_right_key[1] + outline_size,
    )
    top_left_corner = (
        top_left_key[0] - outline_size,
        top_left_key[1] + outline_size - (widen_cutout_around_key_size / 2),
    )

    bottom_row_points = [
        rotate_about_center_of_plane(point, config.angle)
        for point in bottom_row_points
    ]
    top_row_points = [
        rotate_about_center_of_plane(point, config.angle)
        for point in top_row_points
    ]
    bottom_left_corner = rotate_about_center_of_plane(
        bottom_left_corner, config.angle
    )
    bottom_right_corner = rotate_about_center_of_plane(
        bottom_right_corner, config.angle
    )
    top_right_corner = rotate_about_center_of_plane(
        top_right_corner, config.angle
    )
    top_left_corner = rotate_about_center_of_plane(
        top_left_corner, config.angle
    )
    switch_outline_points = [
        bottom_left_corner,
        *bottom_row_points,
        bottom_right_corner,
        top_right_corner,
        *top_row_points,
        top_left_corner,
    ]

    start_of_bottom_row = bottom_row_points[1]

    mirror_base_point = (
        switch_outline_points[-1][0] + (widen_cutout_around_key_size / 2),
        switch_outline_points[-1][1],
    )

    named_points = SimpleNamespace(
        **{
            "start_of_bottom_row": start_of_bottom_row,
            "top_left_corner": top_left_corner,
            "bottom_right_corner": bottom_right_corner,
            "top_right_corner": top_right_corner,
            "mirror_base_point": mirror_base_point,
        }
    )

    return [
        switch_outline_points,
        named_points,
    ]


def calculate_case_outer_points(named_points, outer_frame_size, config):
    bottom_left_corner = find_point_for_angle(
        named_points.start_of_bottom_row,
        -outer_frame_size,
        45 - config.angle,
    )
    bottom_right_corner = find_point_for_angle(
        named_points.bottom_right_corner,
        -outer_frame_size,
        -45 - config.angle,
    )
    top_right_corner = find_point_for_angle(
        named_points.top_right_corner,
        outer_frame_size,
        45 - config.angle,
    )
    return [
        (named_points.top_left_corner[0], bottom_left_corner[1]),
        bottom_left_corner,
        bottom_right_corner,
        top_right_corner,
        (named_points.top_left_corner[0], top_right_corner[1]),
    ]


def calculate_spacer_points(case_outer_points, outer_frame_size, config):
    return [
        case_outer_points[0],
        (
            case_outer_points[0][0],
            find_point_for_angle(
                case_outer_points[1], outer_frame_size, 45 - config.angle
            )[1],
        ),
        find_point_for_angle(
            case_outer_points[1], outer_frame_size, 45 - config.angle
        ),
        find_point_for_angle(
            case_outer_points[2], outer_frame_size, -45 - config.angle
        ),
        find_point_for_angle(
            case_outer_points[3], -outer_frame_size, 45 - config.angle
        ),
        (
            case_outer_points[4][0] + (config.usb_cutout_width / 2),
            find_point_for_angle(
                case_outer_points[3], -outer_frame_size, 45 - config.angle
            )[1],
        ),
        (
            case_outer_points[4][0] + (config.usb_cutout_width / 2),
            case_outer_points[4][1],
        ),
        case_outer_points[3],
        case_outer_points[2],
        case_outer_points[1],
    ]


def calculate_screw_points(spacer_points, outer_frame_size, config):
    screw_distance_from_inner_edge = (
        outer_frame_size - config.inner_frame_size
    ) / 2

    spacer_top_right_corner = spacer_points[4]
    spacer_bottom_left_corner = spacer_points[2]
    spacer_bottom_right_corner = spacer_points[3]

    screw_top_right_corner = find_point_for_angle(
        spacer_top_right_corner,
        screw_distance_from_inner_edge,
        45 - config.angle,
    )

    screw_top_left_corner = (
        config.top_inside_screw_distance_from_usb,
        screw_top_right_corner[1],
    )

    screw_bottom_left_corner = find_point_for_angle(
        spacer_bottom_left_corner,
        -screw_distance_from_inner_edge,
        45 - config.angle,
    )

    screw_bottom_right_corner = find_point_for_angle(
        spacer_bottom_right_corner,
        -screw_distance_from_inner_edge,
        -45 - config.angle,
    )

    return [
        screw_top_left_corner,
        screw_top_right_corner,
        screw_bottom_right_corner,
        screw_bottom_left_corner,
    ]


def calculate_case_geometry(config):
    key_positions = calculate_key_positions(config)

    switch_cutout_points = calculate_switch_cutout_points(
        key_positions,
        config,
    )

    [
        switch_outline_points,
        named_points,
    ] = calculate_switch_outline_points(key_positions, config)

    outer_frame_size = (
        config.outer_frame_size_for_chicago_bolt
        if config.use_chicago_bolt
        else config.outer_frame_size_for_regular_screw
    )

    case_outer_points = calculate_case_outer_points(
        named_points, outer_frame_size, config
    )

    outer_frame_size = (
        config.outer_frame_size_for_chicago_bolt
        if config.use_chicago_bolt
        else config.outer_frame_size_for_regular_screw
    )

    spacer_points = calculate_spacer_points(
        case_outer_points, outer_frame_size, config
    )

    screw_points = calculate_screw_points(
        spacer_points, outer_frame_size, config
    )

    reset_button_radius = 1.5
    reset_button_point = (-30, 101)

    spacer_thickness = (
        config.base_layer_thickness * 2
        if config.has_thicc_spacer
        else config.base_layer_thickness
    )

    return SimpleNamespace(
        **{
            "screws": screw_points,
            "reset_button": SimpleNamespace(
                **{
                    "point": reset_button_point,
                    "radius": reset_button_radius,
                    "thickness": config.base_layer_thickness,
                }
            ),
            "case_outer": case_outer_points,
            "spacer": spacer_points,
            "spacer_thickness": spacer_thickness,
            "thickness": config.base_layer_thickness,
            "switch_outline": switch_outline_points,
            "switch_cutouts": switch_cutout_points,
            "mirror_base_point": named_points.mirror_base_point,
        }
    )


def make_bottom_plate(geometry, config):
    return (
        cq.Workplane()
        .polyline(geometry.case_outer)
        .close()
        .extrude(geometry.thickness)
        .drill_holes(geometry, config)
        .mirror_layer(geometry)
        .drill_reset_button_hole(geometry)
    )


def make_top_plate(geometry, config):
    cutout = (
        cq.Workplane()
        .polyline(geometry.switch_outline)
        .close()
        .extrude(geometry.thickness)
        .mirror_layer(geometry)
    )
    return (
        cq.Workplane()
        .polyline(geometry.case_outer)
        .close()
        .extrude(geometry.thickness)
        .drill_holes(geometry, config)
        .mirror_layer(geometry)
        .cut(cutout)
    )


def make_switch_plate(geometry, config):
    switch_plate = cq.Workplane()

    for switch_cutout in geometry.switch_cutouts:
        switch_plate = switch_plate.polyline(switch_cutout).close()

    return (
        switch_plate.polyline(geometry.case_outer)
        .close()
        .extrude(geometry.thickness)
        .drill_holes(geometry, config)
        .mirror_layer(geometry)
    )


def make_spacer(geometry, config):
    return (
        cq.Workplane()
        .polyline(geometry.spacer)
        .close()
        .extrude(geometry.spacer_thickness)
        .drill_holes(geometry, config)
        .mirror_layer(geometry)
    )


def make_keyboard_parts(user_config={}):
    config = SimpleNamespace(**{**default_config.__dict__, **user_config})

    parts = []

    [time_elapsed, total_time] = timer()

    geometry = calculate_case_geometry(config)
    time_elapsed("Case geometry")

    parts.append(("Top plate", make_top_plate(geometry, config)))
    time_elapsed("Top plate")

    parts.append(("Switch plate", make_switch_plate(geometry, config)))
    time_elapsed("Switch plate")

    if config.has_thicc_spacer:
        parts.append(("Spacer", make_spacer(geometry, config)))
        time_elapsed("Spacer")
    else:
        parts.append(("Spacer 1", make_spacer(geometry, config)))
        parts.append(("Spacer 2", make_spacer(geometry, config)))
        time_elapsed("Spacers")

    parts.append(("Bottom plate", make_bottom_plate(geometry, config)))
    time_elapsed("Bottom plate")

    total_time()

    return parts


if "show_object" in globals():
    keyboard_parts = make_keyboard_parts()

    if not flatten_items:
        keyboard_parts = explode_parts(keyboard_parts, explode_by)

    for layer_name_and_part in keyboard_parts:
        [layer_name, part] = layer_name_and_part
        show_object(part, name=layer_name)
