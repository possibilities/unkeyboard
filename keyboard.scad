use <switches.scad>

module main(view) {
  $fs = 0.1;

  numberOfRows = 2;
  numberOfColumns = 3;

  keyDistanceMm = 19.05;

  topLayerHeightMm = 5;
  middleLayerHeightMm = 3;
  bottomLayerHeightMm = 5;

  // Needs to be 3 or more to ensure wiring doesn't fall off edge. Additional
  // bezel should be added to case.
  switchesBezelMm = 3;

  if (view == "exploded") {
    demoExplodeMm = 20;
    translate([ 0, 0, 2 * demoExplodeMm ]) {
      topLayerSwitches(topLayerHeightMm, switchesBezelMm, numberOfRows,
                       numberOfColumns, keyDistanceMm);
    }
    !translate([ 0, 0, 1 * demoExplodeMm ]) {
      middleLayerSwitches(middleLayerHeightMm, switchesBezelMm, numberOfRows,
                          numberOfColumns, keyDistanceMm);
    }
    translate([ 0, 0, 0 * demoExplodeMm ]) {
      bottomLayerSwitches(bottomLayerHeightMm, switchesBezelMm, numberOfRows,
                          numberOfColumns, keyDistanceMm);
    }
  } else if (view == "assembled") {
    translate([ 0, 0, bottomLayerHeightMm + middleLayerHeightMm ]) {
      topLayerSwitches(topLayerHeightMm, switchesBezelMm, numberOfRows,
                       numberOfColumns, keyDistanceMm);
    }
    !translate([ 0, 0, bottomLayerHeightMm ]) {
      middleLayerSwitches(middleLayerHeightMm, switchesBezelMm, numberOfRows,
                          numberOfColumns, keyDistanceMm);
    }
    translate([ 0, 0, 0 ]) {
      bottomLayerSwitches(bottomLayerHeightMm, switchesBezelMm, numberOfRows,
                          numberOfColumns, keyDistanceMm);
    }
  } else if (view == "top-layer") {
    topLayerSwitches(topLayerHeightMm, switchesBezelMm, numberOfRows,
                     numberOfColumns, keyDistanceMm);
  } else if (view == "middle-layer") {
    middleLayerSwitches(middleLayerHeightMm, switchesBezelMm, numberOfRows,
                        numberOfColumns, keyDistanceMm);
  } else if (view == "bottom-layer") {
    bottomLayerSwitches(bottomLayerHeightMm, switchesBezelMm, numberOfRows,
                        numberOfColumns, keyDistanceMm);
  }
}

view = "assembled";

main(view);
