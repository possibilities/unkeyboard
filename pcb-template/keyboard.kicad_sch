(kicad_sch (version 20211123) (generator eeschema)

  (uuid 9538e4ed-27e6-4c37-b989-9859dc0d49e8)

  (paper "A4")

  (lib_symbols
  )


  (sheet_instances
    (path "/" (page "1"))
  )
)
