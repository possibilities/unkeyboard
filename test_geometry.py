import pytest
from keyboard import presets
from keyboard import calculate_case_geometry_and_make_switch_plate_inner
from types import SimpleNamespace


def build_data_matrix(list_1, list_2):
    data = []
    for item_1 in list_1:
        for item_2 in list_2:
            data.append((item_1, item_2))
    return data


test_data = build_data_matrix(
    presets.__dict__.keys(),
    [
        "usb_rect",
        "screws",
        "reset_button",
        "case_outer",
        "spacer_inner",
        "thickness",
        "spacer_thickness",
    ],
)


@pytest.mark.parametrize("preset_name,geometry_name", test_data)
def test_geometry_usb_rect(preset_name, geometry_name, snapshot):
    preset = presets.__dict__[preset_name]
    [
        geometry,
        switch_plate_inner,
    ] = calculate_case_geometry_and_make_switch_plate_inner(
        SimpleNamespace(**preset)
    )

    assert geometry.__dict__[geometry_name] == snapshot
