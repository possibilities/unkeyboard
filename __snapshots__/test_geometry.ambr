# name: test_geometry_usb_rect[atreus_42-case_outer]
  <class 'list'> [
    <class 'tuple'> (
      -116.32342823957953,
      63.47054789790366,
    ),
    <class 'tuple'> (
      116.32342823957953,
      63.47054789790366,
    ),
    <class 'tuple'> (
      134.60585007869207,
      -40.21421869683712,
    ),
    <class 'tuple'> (
      12.730680709720128,
      -64.75037917065796,
    ),
    <class 'tuple'> (
      -12.730680709720124,
      -64.75037917065796,
    ),
    <class 'tuple'> (
      -134.60585007869207,
      -40.214218696837115,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_42-reset_button]
  <class 'tuple'> (
    -24,
    41,
  )
---
# name: test_geometry_usb_rect[atreus_42-screws]
  <class 'list'> [
    <class 'tuple'> (
      -109.98540861790046,
      54.4189178085103,
    ),
    <class 'tuple'> (
      109.98540861790046,
      54.4189178085103,
    ),
    <class 'tuple'> (
      -11.25,
      54.4189178085103,
    ),
    <class 'tuple'> (
      11.25,
      54.4189178085103,
    ),
    <class 'tuple'> (
      123.3563182982541,
      -36.07410076620267,
    ),
    <class 'tuple'> (
      -123.3563182982541,
      -36.07410076620267,
    ),
    <class 'tuple'> (
      29.335718542082912,
      -55.69874908126461,
    ),
    <class 'tuple'> (
      -29.33571854208291,
      -55.69874908126461,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_42-spacer_inner]
  <class 'list'> [
    <class 'tuple'> (
      -103.64738899622141,
      45.36728771911694,
    ),
    <class 'tuple'> (
      103.64738899622141,
      45.36728771911694,
    ),
    <class 'tuple'> (
      116.50258989990536,
      -27.538179453478996,
    ),
    <class 'tuple'> (
      25.406719953078248,
      -46.647118991871245,
    ),
    <class 'tuple'> (
      -25.406719953078245,
      -46.647118991871245,
    ),
    <class 'tuple'> (
      -116.50258989990536,
      -27.538179453478996,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_42-spacer_thickness]
  3
---
# name: test_geometry_usb_rect[atreus_42-thickness]
  3
---
# name: test_geometry_usb_rect[atreus_42-usb_rect]
  <class 'list'> [
    <class 'tuple'> (
      2.0,
      63.47054789790366,
    ),
    <class 'tuple'> (
      2.0,
      45.36728771911694,
    ),
    <class 'tuple'> (
      -2.0,
      45.36728771911694,
    ),
    <class 'tuple'> (
      -2.0,
      63.47054789790366,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_44-case_outer]
  <class 'list'> [
    <class 'tuple'> (
      -117.88278887502855,
      63.47054789790365,
    ),
    <class 'tuple'> (
      117.88278887502855,
      63.47054789790365,
    ),
    <class 'tuple'> (
      136.1652107141411,
      -40.21421869683712,
    ),
    <class 'tuple'> (
      14.29004134516916,
      -64.75037917065796,
    ),
    <class 'tuple'> (
      -14.290041345169163,
      -64.75037917065796,
    ),
    <class 'tuple'> (
      -136.1652107141411,
      -40.214218696837115,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_44-reset_button]
  <class 'tuple'> (
    -24,
    41,
  )
---
# name: test_geometry_usb_rect[atreus_44-screws]
  <class 'list'> [
    <class 'tuple'> (
      -111.54476925334949,
      54.41891780851029,
    ),
    <class 'tuple'> (
      111.54476925334949,
      54.418917808510294,
    ),
    <class 'tuple'> (
      -11.25,
      54.41891780851029,
    ),
    <class 'tuple'> (
      11.25,
      54.418917808510294,
    ),
    <class 'tuple'> (
      124.91567893370312,
      -36.07410076620267,
    ),
    <class 'tuple'> (
      -124.91567893370312,
      -36.07410076620267,
    ),
    <class 'tuple'> (
      30.895079177531944,
      -55.69874908126461,
    ),
    <class 'tuple'> (
      -30.895079177531947,
      -55.69874908126461,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_44-spacer_inner]
  <class 'list'> [
    <class 'tuple'> (
      -105.20674963167043,
      45.36728771911693,
    ),
    <class 'tuple'> (
      105.20674963167043,
      45.36728771911693,
    ),
    <class 'tuple'> (
      118.06195053535438,
      -27.538179453478996,
    ),
    <class 'tuple'> (
      26.96608058852728,
      -46.647118991871245,
    ),
    <class 'tuple'> (
      -26.966080588527284,
      -46.647118991871245,
    ),
    <class 'tuple'> (
      -118.06195053535438,
      -27.538179453478996,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_44-spacer_thickness]
  3
---
# name: test_geometry_usb_rect[atreus_44-thickness]
  3
---
# name: test_geometry_usb_rect[atreus_44-usb_rect]
  <class 'list'> [
    <class 'tuple'> (
      2.0,
      63.47054789790365,
    ),
    <class 'tuple'> (
      2.0,
      45.36728771911693,
    ),
    <class 'tuple'> (
      -2.0,
      45.36728771911693,
    ),
    <class 'tuple'> (
      -2.0,
      63.47054789790365,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_62-case_outer]
  <class 'list'> [
    <class 'tuple'> (
      -131.7354601711398,
      75.11579487573263,
    ),
    <class 'tuple'> (
      131.73546017113978,
      75.11579487573263,
    ),
    <class 'tuple'> (
      153.31719738592398,
      -47.28031902624011,
    ),
    <class 'tuple'> (
      12.730680709720106,
      -75.11579487573263,
    ),
    <class 'tuple'> (
      -12.730680709720145,
      -75.11579487573263,
    ),
    <class 'tuple'> (
      -153.31719738592403,
      -47.28031902624011,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_62-reset_button]
  <class 'tuple'> (
    -24,
    41,
  )
---
# name: test_geometry_usb_rect[atreus_62-screws]
  <class 'list'> [
    <class 'tuple'> (
      -125.39744054946075,
      66.06416478633928,
    ),
    <class 'tuple'> (
      125.39744054946071,
      66.06416478633928,
    ),
    <class 'tuple'> (
      -11.25,
      66.06416478633928,
    ),
    <class 'tuple'> (
      11.25,
      66.06416478633928,
    ),
    <class 'tuple'> (
      142.067665605486,
      -43.14020109560566,
    ),
    <class 'tuple'> (
      -142.06766560548604,
      -43.14020109560566,
    ),
    <class 'tuple'> (
      29.33571854208289,
      -66.06416478633926,
    ),
    <class 'tuple'> (
      -29.33571854208293,
      -66.06416478633926,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_62-spacer_inner]
  <class 'list'> [
    <class 'tuple'> (
      -119.0594209277817,
      57.01253469694591,
    ),
    <class 'tuple'> (
      119.05942092778166,
      57.01253469694591,
    ),
    <class 'tuple'> (
      135.21393720713726,
      -34.60427978288199,
    ),
    <class 'tuple'> (
      25.406719953078227,
      -57.012534696945906,
    ),
    <class 'tuple'> (
      -25.406719953078266,
      -57.012534696945906,
    ),
    <class 'tuple'> (
      -135.2139372071373,
      -34.60427978288199,
    ),
  ]
---
# name: test_geometry_usb_rect[atreus_62-spacer_thickness]
  3
---
# name: test_geometry_usb_rect[atreus_62-thickness]
  3
---
# name: test_geometry_usb_rect[atreus_62-usb_rect]
  <class 'list'> [
    <class 'tuple'> (
      2.0,
      75.11579487573263,
    ),
    <class 'tuple'> (
      2.0,
      57.01253469694591,
    ),
    <class 'tuple'> (
      -2.0,
      57.01253469694591,
    ),
    <class 'tuple'> (
      -2.0,
      75.11579487573263,
    ),
  ]
---
